### DISCUSSION TITLE
Proper use of swift_allocObject()?

### DISCUSSION THREAD

**Discussion by NSExceptional** [Azoy/Echo#36] [Q&A] [answered]
Given the following extension:
```swift
extension ClassMetadata {
    func createInstance<T: AnyObject>(props: [String: Any] = [:]) -> T {
        var obj = swift_allocObject(
            for: self,
            size: self.classSize,
            alignment: self.instanceAlignmentMask
        )
        
        for (key, value) in props {
            self.set(value: value, forKey: key, on: &obj)
        }
        
        return unsafeBitCast(obj, to: T.self)
    }
}
```

And class definition:
```swift
class JustDeallocateMe {
    var expectation: XCTestExpectation
    
    init(_ expectation: XCTestExpectation) {
        self.expectation = expectation
    }
    
    deinit {
        self.expectation.fulfill()
    }
}
```

I would expect the following test to pass, but it does not:

```swift
func testAllocObject() {
    let expect = XCTestExpectation(description: "deinit")
    let cls = reflectClass(JustDeallocateMe.self)!
    DispatchQueue.global().async {
        var obj: JustDeallocateMe? = cls.createInstance(props: ["expectation": expect])
        obj = nil
    }
    
    self.wait(for: [expect], timeout: 2)
}
```

The `deinit` is never executed. I have tests verifying my `.set(value:forKey:on)` method works. What am I doing wrong?

**Comment by Azoy** [comment, marked as answer]
So the correct way to call `swift_allocObject` is:
```swift
let object = swift_allocObject(
  for: someMetadata,
  size: someMetadata.instanceSize,
  alignment: someMetadata.instanceAlignmentMask
)
```
Where it will return a pointer to a heap object and the fields can be placed in the respective field offset from that pointer.

**Reply by NSExceptional** [reply]
That's what I'm doing (updated to use `instanceAlignmentMask`), but the object seems to be leaking. At least, `deinit` is not called

**Reply by NSExceptional** [reply]
Do I need to manually retain it first? I assume it comes out with +1 and I only need to allow it to be released once it is returned to the caller

Project can be found here if you wanna tinker with the test yourself: https://github.com/NSExceptional/Jsum

**Reply by Azoy** [reply]
Yes, your return should look like: `return Unmanaged.fromOpaque(obj).takeRetainedValue()`. Try that out and let me know. (Also I noticed you use `classSize` instead of `instanceSize` as the size argument.

**Reply by NSExceptional** [reply]
Oh, gotcha. That worked! Man, I wish you could mark sub-replies as the answer down here.

**Reply by NSExceptional** [reply]
Not really related, but do you know how to check if something given as `Any` is an object? Using `thing as AnyObject` will just bridge struct types to `NSValue`…

My test now working has just uncovered a bug in my `set(value:forKey:on:)` method. I need to retain objects when they're being set as `Any`.

**Reply by Azoy** [reply]
You can do something like:
```swift
let box = container(for: someAny)
print(box.type)
print(box.metadata.kind)

if box.metadata.kind == .class || box.metadata.kind == .objcClassWrapper {
  // Any is an object
}
```

**Reply by NSExceptional** [reply]
That works I guess, is there not a way to do it in pure swift though?

**Reply by Azoy** [reply]
~~`thing is AnyObject`~~ this bridges to `NSValue` as well.. What's weird is that the following compiles:
```swift
print(thing is AnyObject) // true
```
but the following does not:
```swift
func test<T: AnyObject>(with: T) {}

test(with: thing) // error: some struct needs to be a class type
```
Difference between the compiler's type system and the runtimes...

**Reply by NSExceptional** [reply]
Yeah, the last like will work if you cast it first, weirdly
